substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/SecuritySubstitutions.java

```diff
@@ -77,102 +79,108 @@
 final class Target_java_security_AccessController {
 
     @Substitute
-    private static <T> T doPrivileged(PrivilegedAction<T> action) throws Throwable {
-        try {
-            return action.run();
-        } catch (Throwable ex) {
-            throw AccessControllerUtil.wrapCheckedException(ex);
-        }
+    public static <T> T doPrivileged(PrivilegedAction<T> action) throws Throwable {
```

Comment:
Please mark as many substitutions as possible as `JDK11OrEarlier`, since JDK 16 already has these methods implemented in Java.

Reply:
Problem is that JDK16's implementation assumes that `getStackAccessControlContext` can easily access method locals to extract `context` and `caller`. Relevant code from Hotspot's `jvm.cpp`:
```
      javaVFrame *priv = vfst.asJavaVFrame();       // executePrivileged
      StackValueCollection* locals = priv->locals();
      StackValue* ctx_sv = locals->at(1); // AccessControlContext context
      StackValue* clr_sv = locals->at(2); // Class<?> caller
```
In order to avoid that - since AFAIK there is no nice and easy way to achieve that - I opted for Espresso approach by implementing parallel stack which is `FastThreadLocal`. Negative side of that approach is that we need to substitute `doPrivileged` even on JDK15+.

Reply:
The parallel stack approach is fine. But still, the `doPrivileged` methods themselves do not need to be substituted anymore on JDK 16.

Reply:
This should be fixed now.